File: flexirule/ruleflow/process/deduplication/deduplication.js
// Copyright (c) 2025, FlexiRule and contributors
frappe.provide("flexirule.processes");

flexirule.processes["Deduplication"] = {
	meta: {
		version: "2.0",
		title: __("Deduplication"),
	},

	get_schema(operation_name, ctx) {
		const operation = this.get_operation(operation_name);
		if (!operation) return [];
		return typeof operation.get_config_fields === "function"
			? operation.get_config_fields(ctx)
			: [];
	},

	get_output_schema(operation_name, config, context) {
		const op_name =
			typeof operation_name === "string"
				? operation_name
				: context?.operation_name || context?.operation || config?.operation;
		const operation = this.get_operation(op_name);
		if (operation && typeof operation.get_output_schema === "function") {
			return operation.get_output_schema(
				typeof operation_name === "string" ? config : operation_name,
				typeof operation_name === "string" ? context : config
			);
		}
		return [];
	},

	operations: [
		{
			func_name: "find_matching_records",
			label: __("Find Matching Records"),
			description: __(
				"Fetch candidate records with blocking filters, score them against the current document, and return structured match details."
			),
			category: "Matching",
			icon: "search",
			color: "#f59e0a",
			validate: (cfg) => {
				if ((!cfg.fields_config || cfg.fields_config.length === 0) && !cfg.fields) {
					return __("Add comparison rules or exact-match fields.");
				}
				return null;
			},
			get_config_fields: (ctx) => [
				{
					fieldtype: "Section Break",
					label: __("Candidate Selection"),
				},
				{
					fieldname: "candidate_doctype",
					fieldtype: "Link",
					options: "DocType",
					label: __("Candidate DocType"),
					description: __("Defaults to the current document DocType if left blank."),
				},
				{
					fieldname: "candidate_limit",
					fieldtype: "Int",
					label: __("Candidate Limit"),
					default: 1000,
				},
				{
					fieldname: "ignore_cancelled",
					fieldtype: "Check",
					label: __("Ignore Cancelled Documents"),
					default: 1,
				},
				{
					fieldname: "exclude_current_document",
					fieldtype: "Check",
					label: __("Exclude Current Document"),
					default: 1,
				},
				{
					fieldname: "stop_after_first_match",
					fieldtype: "Check",
					label: __("Stop After First Match"),
					default: 0,
				},
				{
					fieldtype: "Section Break",
					label: __("Scoring Rules"),
				},
				{
					fieldname: "overall_threshold",
					fieldtype: "Float",
					label: __("Overall Threshold"),
					default: 0.8,
					reqd: 1,
				},
				{
					fieldname: "minimum_fields_matched",
					fieldtype: "Int",
					label: __("Minimum Fields Matched"),
					default: 1,
				},
				{
					fieldname: "fields_config",
					fieldtype: "Table",
					label: __("Field Comparison Rules"),
					reqd: 1,
					fields: get_dedupe_table_fields(),
				},
			],
			get_output_schema: () => [
				{ label: __("Has Match"), value: "has_match", type: "Check" },
				{ label: __("Match Count"), value: "match_count", type: "Int" },
				{ label: __("Best Match"), value: "best_match", type: "JSON" },
				{ label: __("Matches"), value: "matches", type: "JSON" },
				{ label: __("Criteria"), value: "criteria", type: "JSON" },
			],
		},
		{
			func_name: "find_duplicates_in_child_table",
			label: __("Find Child Table Matches"),
			description: __(
				"Find parent documents whose child-table rows contain values already present in the current document."
			),
			category: "Matching",
			icon: "table",
			color: "#0ea5e9",
			validate: (cfg) => {
				if (!cfg.child_table_field) return __("Child Table Field is required.");
				if (!cfg.child_search_field) return __("Child Search Field is required.");
				return null;
			},
			get_config_fields: () => [
				{
					fieldname: "child_table_field",
					fieldtype: "Data",
					label: __("Child Table Field"),
					reqd: 1,
				},
				{
					fieldname: "child_search_field",
					fieldtype: "Data",
					label: __("Field to Compare"),
					reqd: 1,
				},
				{
					fieldname: "normalize_values",
					fieldtype: "Check",
					label: __("Normalize Values"),
					default: 1,
				},
				{
					fieldname: "ignore_cancelled",
					fieldtype: "Check",
					label: __("Ignore Cancelled Documents"),
					default: 1,
				},
				{
					fieldname: "result_limit",
					fieldtype: "Int",
					label: __("Result Limit"),
					default: 1000,
				},
			],
			get_output_schema: () => [
				{ label: __("Has Match"), value: "has_match", type: "Check" },
				{ label: __("Match Count"), value: "match_count", type: "Int" },
				{ label: __("Best Match"), value: "best_match", type: "JSON" },
				{ label: __("Matches"), value: "matches", type: "JSON" },
				{ label: __("Values Checked"), value: "values_checked", type: "JSON" },
			],
		},
	],

	get_operation(name) {
		return this.operations.find((op) => op.func_name === name);
	},
};

const SHARED = {
	validTypes: [
		"Data",
		"Link",
		"Text",
		"Phone",
		"Email",
		"Small Text",
		"Date",
		"Int",
		"Float",
		"Currency",
	],

	async get_field_options(meta) {
		if (!meta?.name) return [];
		const all_fields = await flexirule.utils.get_doctype_fields(meta.name);
		return all_fields
			.filter((f) => SHARED.validTypes.includes(f.fieldtype))
			.map((f) => ({
				label: f.label,
				value: f.value,
				description: f.description,
			}));
	},

	on_algorithm_change(value, ctx) {
		const defaults = {
			Exact: { threshold: 1.0, tolerance: 0, weight: 1.0 },
			Phonetic: { threshold: 0.9, tolerance: 0, weight: 0.5 },
			Fuzzy: { threshold: 0.8, tolerance: 0, weight: 0.5 },
			"Date Distance": { threshold: 0.5, tolerance: 30, weight: 0.4 },
			"Numeric Range": { threshold: 0.5, tolerance: 30, weight: 0.4 },
			Contains: { threshold: 0.8, tolerance: 0, weight: 0.5 },
		};
		const config = defaults[value] || {};
		if (config.threshold !== undefined) ctx.update_field("threshold", config.threshold);
		if (config.tolerance !== undefined) ctx.update_field("tolerance", config.tolerance);
		if (config.weight !== undefined) ctx.update_field("weight", config.weight);
	},
};

function get_dedupe_table_fields() {
	return [
		{
			fieldname: "fieldname",
			fieldtype: "DocField",
			options: "vars.document_type",
			label: __("Field"),
			reqd: 1,
			in_list_view: 1,
			get_options: (cfg, ctx, meta) => SHARED.get_field_options(meta),
		},
		{
			fieldname: "algorithm",
			fieldtype: "Select",
			label: __("Algorithm"),
			options: "Exact\nFuzzy\nPhonetic\nContains\nNumeric Range\nDate Distance",
			reqd: 1,
			default: "Fuzzy",
			in_list_view: 1,
			onchange: (val, row, ctx) => SHARED.on_algorithm_change(val, ctx),
		},
		{
			fieldname: "weight",
			fieldtype: "Float",
			label: __("Weight"),
			default: 0.5,
			in_list_view: 1,
		},
		{
			fieldname: "threshold",
			fieldtype: "Float",
			label: __("Threshold"),
			default: 0.8,
			in_list_view: 1,
			depends_on: "eval:doc.algorithm !== 'Exact'",
		},
		{
			fieldname: "tolerance",
			fieldtype: "Int",
			label: __("Tolerance"),
			default: 30,
			in_list_view: 1,
			depends_on: "eval:in_list(['Numeric Range', 'Date Distance'], doc.algorithm)",
			description: __("Allowed deviation (plus/minus)"),
		},
		{
			fieldname: "normalize",
			fieldtype: "Check",
			label: __("Normalize Values"),
			default: 1,
			in_list_view: 1,
		},
		{
			fieldname: "included_in_filters",
			fieldtype: "Check",
			label: __("Use for Blocking"),
			default: 1,
			in_list_view: 1,
		},
	];
}

File: flexirule/ruleflow/process/deduplication/deduplication.py
# Copyright (c) 2025, FlexiRule and contributors
# For license information, please see license.txt

"""
Deduplication Process

Result-oriented dedupe operations that:
- fetch candidates using blocking filters
- score candidates against the current document
- return structured match payloads for downstream actions
"""

from __future__ import annotations

import frappe
from frappe import _

from flexirule.ruleflow.process.deduplication.scoring import (
	DedupeScoringEngine,
	fetch_candidate_rows,
	normalize_fields_config,
)


def find_matching_records(context, config=None):
	"""Find matching records for the current document and return a structured payload."""
	doc = context.get("doc")
	config = config or {}
	if not doc:
		return _empty_result(config)

	fields_config = normalize_fields_config(config)
	if not fields_config:
		return _empty_result(config)

	candidates = fetch_candidate_rows(doc, fields_config, config)
	engine = DedupeScoringEngine(doc, fields_config, config)
	return engine.find_matches(candidates)


def find_duplicates_in_child_table(context, config=None):
	"""Find parent documents whose child-table values overlap with the current document."""
	doc = context.get("doc")
	config = config or {}
	if not doc:
		return _empty_child_result(config)

	child_field = config.get("child_table_field")
	search_field = config.get("child_search_field")
	if not child_field or not search_field:
		return _empty_child_result(config)

	values = [row.get(search_field) for row in (doc.get(child_field) or []) if row.get(search_field)]
	if not values:
		return _empty_child_result(config)

	if config.get("normalize_values", True):
		values = [DedupeScoringEngine.normalize_value(value) for value in values]

	child_doctype = _get_child_doctype(doc.doctype, child_field)
	if not child_doctype:
		return _empty_child_result(config)

	filters = {
		search_field: ["in", values],
		"parenttype": doc.doctype,
	}
	if config.get("ignore_cancelled", True):
		filters["docstatus"] = ["!=", 2]
	if doc.name and not doc.is_new():
		filters["parent"] = ["!=", doc.name]

	rows = frappe.get_all(
		child_doctype,
		filters=filters,
		fields=["parent", search_field],
		limit=int(config.get("result_limit", 1000) or 1000),
		order_by="modified desc",
	)

	match_map: dict[str, set] = {}
	for row in rows:
		parent = row.get("parent")
		if not parent:
			continue
		value = row.get(search_field)
		if config.get("normalize_values", True):
			value = DedupeScoringEngine.normalize_value(value)
		match_map.setdefault(parent, set()).add(value)

	matches = [
		{
			"name": parent,
			"doctype": doc.doctype,
			"matched_values": sorted(matched_values),
			"score": 100.0,
		}
		for parent, matched_values in match_map.items()
	]
	matches.sort(key=lambda row: (-len(row["matched_values"]), row["name"]))

	return {
		"has_match": bool(matches),
		"match_count": len(matches),
		"matches": matches,
		"best_match": matches[0] if matches else None,
		"values_checked": values,
		"child_table_field": child_field,
		"child_search_field": search_field,
	}


def execute(context, func=None, config=None):
	"""Dispatcher for the deduplication process."""
	if not func:
		frappe.throw(_("Operation function name is required"))

	func = _LEGACY_ALIASES.get(func, func)
	if func not in _OPERATIONS:
		frappe.throw(_("Unknown operation: {0}").format(func))
	return _OPERATIONS[func](context, config or {})


def _empty_result(config=None):
	config = config or {}
	return {
		"has_match": False,
		"match_count": 0,
		"candidate_count": 0,
		"best_match": None,
		"matches": [],
		"criteria": {
			"overall_threshold": float(config.get("overall_threshold", 0.8) or 0.8),
			"minimum_fields_matched": int(config.get("minimum_fields_matched", 1) or 1),
			"fields_config": normalize_fields_config(config),
		},
	}


def _empty_child_result(config=None):
	config = config or {}
	return {
		"has_match": False,
		"match_count": 0,
		"matches": [],
		"best_match": None,
		"values_checked": [],
		"child_table_field": config.get("child_table_field"),
		"child_search_field": config.get("child_search_field"),
	}


def _get_child_doctype(parent_doctype: str, child_field: str) -> str | None:
	for df in frappe.get_meta(parent_doctype).fields:
		if df.fieldname == child_field and df.fieldtype == "Table":
			return df.options
	return None


_OPERATIONS = {
	"find_matching_records": find_matching_records,
	"find_duplicates_in_child_table": find_duplicates_in_child_table,
}

_LEGACY_ALIASES = {
	"find_duplicates": "find_matching_records",
	"find_similar_records": "find_matching_records",
	"find_duplicates_by_fields": "find_matching_records",
	"check_duplicate_and_prevent_save": "find_matching_records",
	"check_similar_and_prevent_save": "find_matching_records",
}
